packages/wrangler/src/pages/dev.ts

```diff
@@ -417,37 +429,126 @@ export const Handler = async (args: PagesDevArguments) => {
+				const workerScriptDirectory = dirname(workerScriptPath);
+				const bundle = await buildRawWorker({
+					workerScriptPath: usingWorkerDirectory
+						? join(workerScriptPath, "index.js")
+						: workerScriptPath,
+					outfile: scriptPath,
+					directory: directory ?? ".",
+					nodejsCompatMode,
+					local: true,
+					sourcemap: true,
+					watch: false,
+					onEnd: () => scriptReadyResolve(),
+					defineNavigatorUserAgent,
+				});
+
+				/*
+				 * EXCLUDE:
+				 *   - "_worker.js" because we're already watching it
+				 *   - everything in "./.wrangler", as it's mostly cache and
+				 *     temporary files
+				 *   - anything outside of the current working directory, since we
+				 *     are expecting `wrangler pages dev` to be run from the Pages
+				 *     project root folder
+				 */
+				const dependencies = Object.keys(bundle.dependencies)
+					.map((dep) => resolve(workerScriptDirectory, dep))
+					.filter(
+						(resolvedDep) =>
+							!resolvedDep.includes(normalize(singleWorkerScriptPath)) &&
+							!resolvedDep.includes(normalize("/.wrangler/")) &&
+							resolvedDep.match(resolve(process.cwd()))
+					);
+
+				// handle non-JS module dependencies, such as wasm/html/binary imports
+				const modules = bundle.modules
+					.filter((module) => !!module.filePath)
+					.map((module) =>
+						resolve(workerScriptDirectory, module.filePath as string)
+					);
+
+				/*
+				 *`bundle.dependencies` and `bundle.modules` combined, will always
+				 * provide us with the most up-to-date list of dependencies we need
+				 * to watch, since they reflect the latest built Worker bundle.
+				 * Therefore, we can safely unwatch all dependencies we have been
+				 * watching so far, and add all the new ones.
+				 */
+				currentBundleDependencies = [...dependencies, ...modules];
+
+				if (watchedBundleDependencies.length) {
+					watcher.unwatch(watchedBundleDependencies);
 				}
+				watcher.add(currentBundleDependencies);
+				watchedBundleDependencies = [...currentBundleDependencies];
 			};
 		}
 
-		await runBuild();
-		watch([workerScriptPath], {
-			persistent: true,
-			ignoreInitial: true,
-		}).on("all", async (event) => {
-			if (event === "unlink") {
-				return;
-			}
+		/*
+		 * Improve developer experience by debouncing the re-building
+		 * of the Worker script.
+		 *
+		 * "Debouncing ensures that exactly one signal is sent for an
+		 * event that may be happening several times — or even several
+		 * hundreds of times over an extended period. As long as the
+		 * events are occurring fast enough to happen at least once in
+		 * every detection period, the signal will not be sent!"
+		 * (http://unscriptable.com/2009/03/20/debouncing-javascript-methods/)
+		 *
+		 * This is great for the typing experience for example, when ideally
+		 * we want to re-build the Worker script once, when the user is done
+		 * typing, as opposed to once per each typed character.
```

Comment:
Why would this otherwise build once per typed character? Surely this would only trigger on file save?

Reply:
indeed, that was not very accurate of me to say. I'll rephrase

Reply:
✅